### PR TITLE
feat: add support _EXPERIMENTAL_DAGGER_CLI_BIN to rust

### PR DESCRIPTION
**Pull request by wingyplus**
From the root of the rust sdk, run:

    $ env _EXPERIMENTAL_DAGGER_CLI_BIN=../../bin/dagger cargo run --example first-pipeline

The sdk will use dagger binary defined in `_EXPERIMENTAL_DAGGER_CLI_BIN` to run the pipeline. And modify the engine to try running on local cli when session port & token doesn't set, if this configuration is not set, it will fallback to download cli method.

And also fix tests that failure because of using deleted apis. And re-generate code to make lint happy.

### CODE REVIEW

**Review by sipsma — APPROVED** [review]
LGTM, thanks!

**Comment by sipsma** [comment]
I'm a rust noob so would prefer to get @kjuulh signoff before merge

File: sdk/rust/crates/dagger-sdk/src/core/engine.rs
Comment (kjuulh) [inline sdk/rust/crates/dagger-sdk/src/core/engine.rs:56]:
```suggestion
    ) -> eyre::Result<Option<(ConnectParams, tokio::process::Child)>> {
```

**Review by kjuulh — CHANGES_REQUESTED** [review]
Some small comments. Feel free to ping me if you want to discuss the changes =D

We may have to duplicate the ENGINE version so that we've got both a 

- Fetched engine version 
- Running engine version

This may be useful because if we're not using the fetcher, and instead the one provided by dagger run, or just a binary, then we don't control what engine we're running on.

Otherwise, we could also just add a flag set in the user agent, such that we can see that we're running under another engine. Or use an external binary.

File: sdk/rust/crates/dagger-sdk/src/gen.rs
Comment (wingyplus) [reply sdk/rust/crates/dagger-sdk/src/gen.rs:172]: I generate it via `./hack/make sdk::rust`. So the generate task would do something wrong?

Comment (wingyplus) [reply sdk/rust/crates/dagger-sdk/src/gen.rs:172]: Oh. I forgot that we need `./hack/make sdk:rust:bump` to bump the engine version. 😂

Comment (wingyplus) [reply sdk/rust/crates/dagger-sdk/src/gen.rs:172]: What if we bump it to `0.6.4` like Go does?

File: sdk/rust/crates/dagger-sdk/src/core/engine.rs
Comment (wingyplus) [reply sdk/rust/crates/dagger-sdk/src/core/engine.rs:36]: If we put `?` after await then it would raise with error:

```
Error: failed to connect to dagger engine

Caused by:
    environment variable not found

Location:
    crates/dagger-sdk/examples/first-pipeline/main.rs:3:18
```

The error above come from running `cargo run --example first-pipeline`. It probably because of I used `await?` during fetch env variable. What should I suppose to fix in this case?

**Comment by wingyplus** [comment]
> Some small comments. Feel free to ping me if you want to discuss the changes =D
> 
> We may have to duplicate the ENGINE version so that we've got both a
> 
>     * Fetched engine version
> 
>     * Running engine version
> 
> 
> This may be useful because if we're not using the fetcher, and instead the one provided by dagger run, or just a binary, then we don't control what engine we're running on.
> 
> Otherwise, we could also just add a flag set in the user agent, such that we can see that we're running under another engine. Or use an external binary.

I'm not sure if this related to https://github.com/dagger/dagger/pull/5315? So we could ensure that the rust SDK still compat with the current engine that running on.

File: sdk/rust/crates/dagger-sdk/src/gen.rs
Comment (kjuulh) [reply sdk/rust/crates/dagger-sdk/src/gen.rs:172]: Yep. that is fine as well.

A bump should be sufficient

File: sdk/rust/crates/dagger-sdk/src/core/engine.rs
Comment (kjuulh) [reply sdk/rust/crates/dagger-sdk/src/core/engine.rs:36]: You can just return a None instead then.

We don't care about the error, so we just return an Option, 

If you don't need the error at all.

It can also be done like so.

```rust
if let Some((conn, child)) = self.from_local_cli(cfg).await.ok() { //-> this will transform an Err into an Option. 
```

**Comment by kjuulh** [comment]
> > Some small comments. Feel free to ping me if you want to discuss the changes =D
> > We may have to duplicate the ENGINE version so that we've got both a
> > ```
> > * Fetched engine version
> > 
> > * Running engine version
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This may be useful because if we're not using the fetcher, and instead the one provided by dagger run, or just a binary, then we don't control what engine we're running on.
> > Otherwise, we could also just add a flag set in the user agent, such that we can see that we're running under another engine. Or use an external binary.
> 
> I'm not sure if this related to #5315? So we could ensure that the rust SDK still compat with the current engine that running on.

There is a bit of overlap for sure. 

But I think it is fine with going what we've got now, as long as just get the bump fixed, so that the generated code matches the version it is running under.

Lets keep in mind that we may need the compatibility checks at some point =D

---

All in all, to get the issues fixed, we just need to trigger bump, so that we run under the right version, and get the function fixed, so that it either return an Error, which we transform with `.ok` or returning an Option inside the function. Up to you, I don't mind either way

Hopefully the direction to resolution isn't too unclear =D

**Comment by wingyplus** [comment]
@kjuulh All issues are addressed. Please kindly review it again. 🙇

**Review by kjuulh — APPROVED** [review]
Looks good

**Pull request merged**
